include/pagmo/algorithms/pso.hpp

```diff
@@ -0,0 +1,875 @@
+                    if (new_x < lb[d]) {
+                        new_x = lb[d];
+                        m_V[p][d] = 0.;
+                        //					new_x = boost::uniform_real<double>(lb[d],ub[d])(m_drng);
+                        //					V[p][d] = new_x - X[p][d];
+                    } else if (new_x > ub[d]) {
+                        new_x = ub[d];
+                        m_V[p][d] = 0.;
+                        //					new_x = boost::uniform_real<double>(lb[d],ub[d])(m_drng);
+                        //					V[p][d] = new_x - X[p][d];
+                    }
+                    X[p][d] = new_x;
+                }
+                // We evaluate here the new individual fitness
+                // as to be able to update the global best in real time
+                fit[p] = prob.fitness(X[p]);
+
+                if (fit[p] <= lbfit[p]) {
+                    // update the particle's previous best position
+                    lbfit[p] = fit[p];
+                    lbX[p] = X[p];
+                    // update the best position observed so far by any particle in the swarm
+                    // (only performed if swarm topology is gbest)
+                    if ((m_neighb_type == 1u || m_neighb_type == 4u) && (fit[p] <= best_fit)) {
+                        best_neighb = X[p];
+                        best_fit = fit[p];
+                        best_fit_improved = true;
+                    }
+                }
+            } // End of loop on the population members
+            // reset swarm topology if no improvement was observed in the best found fitness value
+            if (m_neighb_type == 4u && !best_fit_improved) initialize_topology__adaptive_random(neighb);
+            // Logs and prints (verbosity modes > 1: a line is added every m_verbosity generations)
+            if (m_verbosity > 0u) {
+                // Every m_verbosity generations print a log line
+                if (gen % m_verbosity == 1u || m_verbosity == 1u) {
+                    // We compute the number of function evaluations made
+                    auto feval_count = prob.get_fevals() - fevals0;
+                    // We compute the average across the swarm of the best fitness encountered
+                    vector_double local_fits(swarm_size);
+                    for (decltype(swarm_size) i = 0u; i < swarm_size; ++i) {
+                        local_fits[i] = lbfit[i][0];
+                    }
+                    auto lb_avg = std::accumulate(local_fits.begin(), local_fits.end(), 0.)
+                                  / static_cast<double>(local_fits.size());
+                    // We compute the best fitness encounterd so far across generations and across the swarm
+                    auto idx_best = std::distance(std::begin(local_fits),
+                                                  std::min_element(std::begin(local_fits), std::end(local_fits)));
+                    auto best = local_fits[static_cast<unsigned int>(idx_best)];
```

Comment:
To be really pedantic, there's 2 things potentially problematic:

* ``std::distance`` will return a signed quantity (that's the reason to do the ``static_cast`` I presume), and this means that it could overflow the size type of ``local_fits``;
* the cast to ``unsigned`` should probably be a cast to the size type of ``local_fits``.

Unfortunately, I don't have much to offer in terms of constructive suggestions, so this code can go in without modifications as far as I am concerned. Properly fixing this would require some ugly checks, maybe you could add a comment as a FIXME or TODO for the future?